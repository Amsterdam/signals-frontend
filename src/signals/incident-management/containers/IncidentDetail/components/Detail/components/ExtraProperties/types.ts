export type Text = string;

export interface Checkbox {
  label: string;
  value: boolean;
}

export interface Radio {
  id: string;
  label: string;
  info: string;
}

export interface Container {
  description: string;
  type: string;
  id: string;
}

export type StreetLightId = string;

export type Answer = Text | Checkbox | Radio | (Container | StreetLightId)[];

export interface Item {
  id: string;
  label: string;
  answer: Answer;
  category_url: string;
}

type LegacyAnswer = Text | Checkbox | Radio;

export type LegacyItem = Record<string, LegacyAnswer>;

export interface MappedLegacyItem extends Pick<Item, 'id' | 'label'> {
  answer: LegacyAnswer;
}

