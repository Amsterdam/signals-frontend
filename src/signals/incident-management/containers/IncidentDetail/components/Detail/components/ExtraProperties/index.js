import React, { Fragment } from 'react';
import isObject from 'lodash.isobject';
import isArray from 'lodash.isarray';
import isBoolean from 'lodash.isboolean';

import { extraPropertiesType } from 'shared/types';

const getValue = answer => {
  if (isArray(answer)) {
    return answer.map(item => isObject(item) ? item.label : item).join(', ');
  }
  if (isObject(answer)) {
    if (isBoolean(answer.value)) {
      return answer.value ? answer.label : 'Nee';
    }
    return answer.label;
  }

  return answer;
};

const ExtraProperties = ({ items }) =>
  items &&
  items.map(item => (
    <Fragment key={item.id}>
      <dt data-testid="extra-properties-definition">{item.label}</dt>
      <dd data-testid="extra-properties-value">{getValue(item.answer)}</dd>
    </Fragment>
  ));

ExtraProperties.defaultProps = {
  items: [],
};

ExtraProperties.propTypes = {
  items: extraPropertiesType,
};

export default ExtraProperties;
