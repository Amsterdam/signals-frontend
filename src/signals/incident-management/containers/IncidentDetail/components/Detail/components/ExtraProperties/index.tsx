// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2021 -2022 Gemeente Amsterdam
import { Fragment } from 'react'
import type { FunctionComponent } from 'react'

import { mapExtraPropertiesToJSX } from 'shared/services/map-extra-properties'
import type {
  MappedLegacyItem,
  Item,
  ExtraPropertiesTypes,
} from 'shared/types/extraProperties'

interface ExtraPropertiesProps {
  items?: ExtraPropertiesTypes
}

const ExtraProperties: FunctionComponent<ExtraPropertiesProps> = ({
  items = [],
}) => {
  // Some incidents have been stored with values for their extra properties that is incompatible with the current API
  // We therefore need to check if we're getting an array or an object
  const itemList: (Item | MappedLegacyItem)[] = Array.isArray(items)
    ? items
    : Object.entries(items).map(([question, answer], index) => ({
        label: question,
        answer,
        id: index.toString(),
      }))

  return (
    <Fragment>
      {itemList.map((item) => (
        <Fragment key={item.id}>
          <dt data-testid="extra-properties-definition">{item.label}</dt>
          <dd data-testid="extra-properties-value">
            {mapExtraPropertiesToJSX(item.answer)}
          </dd>
        </Fragment>
      ))}
    </Fragment>
  )
}

export default ExtraProperties
