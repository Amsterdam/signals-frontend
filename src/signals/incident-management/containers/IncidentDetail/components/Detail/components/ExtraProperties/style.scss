@import "config.scss";

.extra-properties {
  padding-bottom: 17px;
}