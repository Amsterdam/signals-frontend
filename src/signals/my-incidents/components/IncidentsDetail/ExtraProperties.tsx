// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { Fragment } from 'react'

import { mapExtraPropertiesToJSX } from 'shared/services/map-extra-properties'
import type {
  ExtraPropertiesTypes,
  MappedLegacyItem,
} from 'shared/types/extraProperties'

import { FormTitle, StyledDD } from './styled'

interface Props {
  items?: ExtraPropertiesTypes
}

export const ExtraProperties = ({ items }: Props) => {
  if (!items) return null

  const itemList: ExtraPropertiesTypes | MappedLegacyItem[] = Array.isArray(
    items
  )
    ? items
    : Object.entries(items).map(([question, answer], index) => ({
        label: question,
        answer,
        id: index.toString(),
      }))

  return (
    <dl>
      {itemList.map((item) => (
        <Fragment key={item.id}>
          <FormTitle>{item.label}</FormTitle>
          <StyledDD>{mapExtraPropertiesToJSX(item.answer)}</StyledDD>
        </Fragment>
      ))}
    </dl>
  )
}
