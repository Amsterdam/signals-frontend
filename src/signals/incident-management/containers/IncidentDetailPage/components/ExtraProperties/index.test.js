import React from 'react';
import { shallow } from 'enzyme';

import ExtraProperties from './index';

describe('<ExtraProperties />', () => {
  let wrapper;
  const props = {
    items: [
      {
        id: 'extra_straatverlichting',
        label: 'Gaat uw melding over één of over meer lampen?',
        category_url: '/signals/v1/public/terms/categories/wegen-verkeer-straatmeubilair/sub_categories/straatverlichting-openbare-klok',
        answer: {
          id: 'meer_lampen',
          label: 'Meer lampen'
        }
      },
      {
        id: 'extra_straatverlichting_wat',
        label: 'Wat is er aan de hand met de lamp(en)?',
        category_url: '/signals/v1/public/terms/categories/wegen-verkeer-straatmeubilair/sub_categories/straatverlichting-openbare-klok',
        answer: [{ id: 'brandt_niet', label: 'Brandt niet' }, { id: 'brandt_overdag', label: 'Brandt overdag' }]
      },
      {
        id: 'extra_straatverlichting_nummer',
        label: 'Hebt u een nummer van (één van) de lamp(en)?',
        category_url: '/signals/v1/public/terms/categories/wegen-verkeer-straatmeubilair/sub_categories/straatverlichting-openbare-klok',
        answer: '42'
      },
      {
        id: 'extra_bedrijven_vaker',
        label: 'Gebeurt het vaker?',
        category_url: '/signals/v1/public/terms/categories/wegen-verkeer-straatmeubilair/sub_categories/straatverlichting-openbare-klok',
        answer: { label: 'Ja, het gebeurt vaker', value: false }
      },
      {
        id: 'extra_bedrijven_gezien',
        label: 'Heeft u het gezien?',
        category_url: '/signals/v1/public/terms/categories/wegen-verkeer-straatmeubilair/sub_categories/straatverlichting-openbare-klok',
        answer: { label: 'Ja, het gebeurt vaker', value: true }
      },
      {
        id: 'lampen',
        label: 'Welke lampen?',
        category_url: '/signals/v1/public/terms/categories/wegen-verkeer-straatmeubilair/sub_categories/straatverlichting-openbare-klok',
        answer: [126543, 87654]
      }
    ]
  };

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    wrapper = shallow(<ExtraProperties {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly without items', () => {
    wrapper = shallow(<ExtraProperties />);

    expect(wrapper).toMatchSnapshot();
  });
});
