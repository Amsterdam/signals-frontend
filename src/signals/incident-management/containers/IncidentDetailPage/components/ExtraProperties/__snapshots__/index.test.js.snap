// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExtraProperties /> should render correctly 1`] = `
<div
  className="extra-properties"
>
  <dl
    key="extra_straatverlichting"
  >
    <dt
      className="incident-detail__body__definition"
    >
      Gaat uw melding over één of over meer lampen?
    </dt>
    <dd
      className="incident-detail__body__value"
    >
      Meer lampen
    </dd>
  </dl>
  <dl
    key="extra_straatverlichting_wat"
  >
    <dt
      className="incident-detail__body__definition"
    >
      Wat is er aan de hand met de lamp(en)?
    </dt>
    <dd
      className="incident-detail__body__value"
    >
      Brandt niet, Brandt overdag
    </dd>
  </dl>
  <dl
    key="extra_straatverlichting_nummer"
  >
    <dt
      className="incident-detail__body__definition"
    >
      Hebt u een nummer van (één van) de lamp(en)?
    </dt>
    <dd
      className="incident-detail__body__value"
    >
      42
    </dd>
  </dl>
  <dl
    key="extra_bedrijven_vaker"
  >
    <dt
      className="incident-detail__body__definition"
    >
      Gebeurt het vaker?
    </dt>
    <dd
      className="incident-detail__body__value"
    >
      Nee
    </dd>
  </dl>
  <dl
    key="extra_bedrijven_gezien"
  >
    <dt
      className="incident-detail__body__definition"
    >
      Heeft u het gezien?
    </dt>
    <dd
      className="incident-detail__body__value"
    >
      Ja, het gebeurt vaker
    </dd>
  </dl>
</div>
`;

exports[`<ExtraProperties /> should render correctly without items 1`] = `
<div
  className="extra-properties"
/>
`;
